File: AI/Python/Astar_Bard.py
class Node:

  def __init__(self, name):
    self.name = name
    self.g = float("inf")
    self.h = float("inf")
    self.parent = None

  def __lt__(self, other):
    return self.f() < other.f()

  def f(self):
    return self.g + self.h
  
def a_star(graph, start, goal, heuristic):
  """
  Recherche le chemin le plus court entre un nœud initial et un nœud final dans un graphe.

  Args:
    graph: Le graphe à parcourir.
    start: Le nœud initial.
    goal: Le nœud final.
    heuristic: Une fonction d'évaluation qui estime la distance restante à parcourir jusqu'à l'arrivée.

  Returns:
    Le chemin le plus court entre les deux nœuds.
  """

  # Initialisation
  open_set = [(0, Node(start))]
  closed_set = []
  came_from = {}

  # Boucle principale
  while open_set:
    # Extraction du nœud le plus prometteur
    _, node = open_set.pop(0)

    # Si le nœud est l'arrivée, on a trouvé le chemin
    if node == goal:
      return reconstruct_path(came_from, goal)

    # Ajout du nœud à la liste des nœuds visités
    closed_set.append(node)

    # Exploration des voisins
    for neighbor, cost in graph[node].items():
      # Si le voisin n'est pas encore visité
      if neighbor not in closed_set:
        # Calcul de la fonction d'évaluation du voisin
        new_g = node.g + cost

        # Si le voisin n'est pas encore dans la liste des nœuds à explorer
        if neighbor not in open_set or new_g < open_set[neighbor][0]:
          # Mise à jour de la fonction d'évaluation du voisin
          open_set.append((new_g, Node(neighbor)))

          # Mise à jour de la distance restante estimée
          neighbor.h = heuristic(neighbor)

          # Ajout du voisin à la liste des nœuds à explorer
          came_from[neighbor] = node

  # Si on arrive ici, c'est que le chemin n'existe pas
  return None


def reconstruct_path(came_from, end):
  """
  Reconstruct le chemin entre un nœud final et le nœud initial.

  Args:
    came_from: Un dictionnaire qui mappe chaque nœud à son prédécesseur.
    end: Le nœud final.

  Returns:
    Le chemin entre les deux nœuds.
  """

  path = [end]
  while end in came_from:
    end = came_from[end]
    path.append(end)
  return path[::-1]


# Définition du graphe
graph = {
  "A": {"B": [2, 3], "C": [10, 5], "D": [3, 6]},
  "B": {"A": [2, 9], "E": [8, 8]},
  "C": {"A": [10, 9], "D": [2, 6], "G": [2, 2]},
  "D": {"A": [3, 9], "C": [2, 5], "F": [4, 4]},
  "E": {"B": [8, 8], "F": [5, 4], "H": [10, 2]},
  "F": {"D": [4, 6], "E": [5, 8], "G": [5, 2]},
  "G": {"C": [2, 5], "F": [5, 4], "H": [1, 0]},
  "H": {"E": [10, 8], "G": [1, 2]}
}

# Définition des nœuds initial et final
start = "A"
goal = "H"

# Définition de l'heuristique
def heuristic(node):
  return abs(node.x - goal.x) + abs(node.y - goal.y)

# Recherche du chemin
path = a_star(graph, start, goal, heuristic)

File: AI/CPP/Astar_Bard.cpp
#include <iostream>
#include <set>
#include <map>
#include <set>
#include <cmath>

struct Node {
  std::string name;
  int g;
  int h;
  Node* parent;
  Node(std::string start) : name(start), g(0), h(heuristic(new Node(start))), parent(nullptr) {}
  Node(std::string name, int g, int h, Node* parent) : name(name), g(g), h(h), parent(parent) {}
};

std::set<Node*> open_set;
std::set<Node*> closed_set;
std::map<Node*, Node*> came_from;

std::string start = "A";
std::string goal = "C";

int heuristic(Node* node) {
  // Calcul de la distance euclidienne entre le nœud et la destination
  int dx = node->name[0] - goal[0];
  int dy = node->name[1] - goal[1];
  return sqrt(dx * dx + dy * dy);
}

Node* a_star(std::map<std::string, std::map<std::string, int>>& graph, std::string start, std::string goal) {
  open_set.insert(new Node(start, 0, heuristic(new Node(start)), nullptr));

  while (!open_set.empty()) {
    Node* current = *(open_set.begin());
    open_set.erase(current);
    closed_set.insert(current);

    if (current->name == goal) {
      return current;
    }

    for (const auto& neighbor : graph[current->name]) {
      Node* neighbor_node = new Node(neighbor.first, current->g + neighbor.second, heuristic(new Node(neighbor.first)), current);

      // Vérifie si le nœud voisin n'a pas déjà été exploré
      if (closed_set.find(neighbor_node) == closed_set.end()) {
        // Vérifie si le nœud voisin est plus court que le nœud le plus court connu
        if (open_set.find(neighbor_node) == open_set.end() || neighbor_node->g < (*open_set.find(neighbor_node))->g) {
          // Ajoute le nœud voisin à la liste des nœuds à explorer
          open_set.insert(neighbor_node);
          // Conserve l'historique des parents des nœuds
          came_from[neighbor_node] = current;
        }
      }
    }
  }

  return nullptr;
}

int main() {
  std::map<std::string, std::map<std::string, int>> graph = {
    {"A", {{"B", 2}, {"C", 3}}},
    {"B", {{"A", 1}, {"C", 2}}},
    {"C", {{"A", 3}, {"B", 1}}},
  };


  Node* path = a_star(graph, start, goal);
  if (path == nullptr) {
    std::cout << "Aucun chemin trouvé" << std::endl;
  } else {
    std::cout << "Chemin trouvé :" << std::endl;
    while (path != nullptr) {
      std::cout << path->name << " ";
      path = path->parent;
    }
    std::cout << std::endl;
  }

  return 0;
}